Add support for Memory64 in WebAssembly in IPInt Tier

colelao opened the pull request:
#### 26868df946342191697031000d7c69d062e29193
<pre>
Add support for Memory64 in WebAssembly in IPInt Tier
<a href="https://bugs.webkit.org/show_bug.cgi?id=300539">https://bugs.webkit.org/show_bug.cgi?id=300539</a>
<a href="https://rdar.apple.com/162406385">rdar://162406385</a>

Reviewed by Keith Miller.

This patch allows the WasmFunctionParser to accept 64 bit pointers while performing
memory operations. This includes all the load and store operations, the bulk memory
operations, and the size and grow operations.

Test: JSTests/wasm/stress/load-and-store-with-64-bit-address.js

* JSTests/wasm/stress/memory64-bulk-memory.js: Added.
(async test):
* JSTests/wasm/stress/memory64-grow-and-size.js: Added.
(async test):
(async testGrowByZero):
(async testNoMaximum):
(async testLargeGrowValue):
* JSTests/wasm/stress/memory64-load-and-store.js: Added.
(getWasmTypeWidth):
(let.wat.module):
(join):
(async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt64AndLength):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::load):
(JSC::Wasm::FunctionParser&lt;Context&gt;::store):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::load):
(JSC::Wasm::IPIntGenerator::store):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::init):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/JavaScriptCore/wasm/WasmMemoryInformation.cpp:
(JSC::Wasm::MemoryInformation::MemoryInformation):
* Source/JavaScriptCore/wasm/WasmMemoryInformation.h:
(JSC::Wasm::MemoryInformation::isMemory64 const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::memoryInit):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::limitsFlagIsValid):
(JSC::Wasm::SectionParser::parseResizableLimits):
(JSC::Wasm::SectionParser::parseTableHelper):
(JSC::Wasm::SectionParser::parseMemoryHelper):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::memoryInit):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:

Canonical link: <a href="https://commits.webkit.org/304075@main">https://commits.webkit.org/304075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70417404e95304661ef5c11fb128de2b8118032b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134293 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45514 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141868 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86329 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102681 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69945 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120373 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83474 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5018 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2632 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2626 "Build is being retried. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Failed to compile WebKit; Failed to compile WebKit; Unable to build WebKit without PR, retrying build (failure)") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126372 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114298 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144520 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132809 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6469 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39089 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111074 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6553 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5442 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111328 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28271 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4854 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116642 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60257 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6521 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34853 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165766 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6356 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70076 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43316 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6593 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6469 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
Note that memory64 requires a lot of compiler work.

1. You need to make wasm bound checking compiler code work for memory64
2. Bound checking code has various optimizations in B3 compiler as well. So you need to update and/or add new optimizations to allow B3 compiler to remove these checks when possible.
3. You need to ensure that fast memory is not used for memory64 since page fault based handling does not work
4. You need to change wasm address calculation code for memory64 in all tiers since they have strong assumption that input is 32bit

Constellation reviewed (commented):
We need to disable this via Options since it is not having BBQ and OMG.

Andrew-Fryer commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -938,246 +938,232 @@ macro ipintCheckMemoryBound(mem, scratch, size)
     # pop index
     popMemoryIndex(t0, t2)
-    loadi IPInt::Const32Metadata::value[MC], t2
+		loadMemoryOffsetAndAdvanceMC(t2, t3, t4)
```
indentation looks wonky; maybe tabs vs. spaces IDE thing?

colelao replied:
Good catch, thanks

ast-hugger commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -658,6 +658,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, useWasmSIMD, true, Normal, "Allow the new simd instructions and types from the wasm simd spec."_s) \
     v(Bool, useWasmRelaxedSIMD, false, Normal, "Allow the relaxed simd instructions and types from the wasm relaxed simd spec."_s) \
     v(Bool, useWasmTailCalls, true, Normal, "Allow the new instructions from the wasm tail calls spec."_s) \
+    v(Bool, useWasmMemory64, false, Normal, "Allow the Memory64 proposal for WebAssembly. This feature is currently only supported in the IPInt tier."_s) \
```
Drive-by comment: let's keep these sorted alphabetically for ease of future maintenance.

colelao replied:
Done 'n done!

ast-hugger commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -547,12 +547,12 @@ WASM_IPINT_EXTERN_CPP_DECL(table_grow, IPIntStackEntry* sp, TableGrowMetadata* m
 }
 
-WASM_IPINT_EXTERN_CPP_DECL(memory_init, int32_t dataIndex, IPIntStackEntry* sp)
+WASM_IPINT_EXTERN_CPP_DECL(memory_init, int64_t dataIndex, IPIntStackEntry* sp)
```
Does `dataIndex` need to be widened to `int64_t`? The proposal maintains `dataidx` as a `u32`, and downstream logic in `Wasm::memoryInit()` is unchanged.

colelao replied:
You're right. I updated it!

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -938,246 +938,232 @@ macro ipintCheckMemoryBound(mem, scratch, size)
+	loadb JSWebAssemblyInstance::m_cachedIsMemory64[wasmInstance], tmpReg
+	btiz tmpReg, .memory32
+	loadq IPInt::Const64Metadata::value[MC], dstReg
+  loadb IPInt::Const64Metadata::instructionLength[MC], instrLenReg
```
Spacing seems off on this line?

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -938,246 +938,232 @@ macro ipintCheckMemoryBound(mem, scratch, size)
+	jmp .done
+.memory32:
+	loadi IPInt::Const32Metadata::value[MC], dstReg
+  loadb IPInt::Const32Metadata::instructionLength[MC], instrLenReg
```
ditto

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -297,10 +326,11 @@ auto SectionParser::parseTableHelper(bool isImport) -> PartialResult
     bool isShared = false;
-    PartialResult limits = parseResizableLimits(initial, maximum, isShared, LimitsType::Table);
+    bool isTable64;
+    PartialResult limits = parseResizableLimits<LimitsType::Table>(initial, maximum, isShared, isTable64);
```
Can we add an `ASSERT(!isShared);`

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -354,13 +384,15 @@ auto SectionParser::parseMemoryHelper(bool isImport) -> PartialResult
+        PartialResult limits = parseResizableLimits<LimitsType::Memory>(initial, maximum, isShared, isMemory64);
         if (!limits) [[unlikely]]
             return makeUnexpected(WTFMove(limits.error()));
+        WASM_PARSER_FAIL_IF(isShared && !maximum, "shared memory must have a maximum size"_s);
```
We should keep this in `parseResizableLimits` since when wasm inevitably adds shared tables this requirement will be required there too.

kmiller68 commented on JSTests/wasm/memory64/bulk-memory.js:
```diff
@@ -0,0 +1,63 @@
+//@ skip if $addressBits <= 32
+//@ runDefaultWasm("-m", "--useJIT=0", "--useWasmMemory64=1")
```
Nit: These could be `--useBBQJIT=0` for slightly more coverage.

kmiller68 reviewed (approved):
r=me with some comments.

colelao committed:
Add support for Memory64 in WebAssembly in IPInt Tier
https://bugs.webkit.org/show_bug.cgi?id=300539
rdar://162406385

Reviewed by Keith Miller.

This patch allows the WasmFunctionParser to accept 64 bit pointers while performing
memory operations. This includes all the load and store operations, the bulk memory
operations, and the size and grow operations.

Test: JSTests/wasm/stress/load-and-store-with-64-bit-address.js

* JSTests/wasm/stress/memory64-bulk-memory.js: Added.
(async test):
* JSTests/wasm/stress/memory64-grow-and-size.js: Added.
(async test):
(async testGrowByZero):
(async testNoMaximum):
(async testLargeGrowValue):
* JSTests/wasm/stress/memory64-load-and-store.js: Added.
(getWasmTypeWidth):
(let.wat.module):
(join):
(async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addLEB128ConstantInt64AndLength):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::load):
(JSC::Wasm::FunctionParser<Context>::store):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::load):
(JSC::Wasm::IPIntGenerator::store):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::init):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/JavaScriptCore/wasm/WasmMemoryInformation.cpp:
(JSC::Wasm::MemoryInformation::MemoryInformation):
* Source/JavaScriptCore/wasm/WasmMemoryInformation.h:
(JSC::Wasm::MemoryInformation::isMemory64 const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::memoryInit):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::limitsFlagIsValid):
(JSC::Wasm::SectionParser::parseResizableLimits):
(JSC::Wasm::SectionParser::parseTableHelper):
(JSC::Wasm::SectionParser::parseMemoryHelper):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::memoryInit):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:

Canonical link: https://commits.webkit.org/304075@main
